### PR TITLE
Add missing validation for nested collections

### PR DESCRIPTION
**Pull request by simonedelmann**
Adds missing validation for nested collections (fixes https://github.com/vapor/vapor/issues/2158):
`addForEach(_ key: ValidationKey, required: Bool = true, _ nested: (inout Validations) -> ())`

```swift
// Assuming this model
struct User {
    struct Tag {
        var title: String
    }
    var tags: [Tag] 
}

// Nested validations
extension User: Validatable {
    static func validations(_ v: inout Validations) {
        v.add("tags", as: [Tag].self)
        v.add(forEach: "tags") { tag in 
            tag.add("title", 
                as: String.self, 
                is: .count(5...) && .characterSet(.alphanumerics + .whitespaces)
        }
    }
}
```
Errors are thrown like this:
```
tags Index 0 title contains '€' (allowed: whitespace, A-Z, a-z, 0-9) and Index 1 title is less than minimum of 5 character(s)
```

### CODE REVIEW

File: Sources/Vapor/Validation/ValidatorResult.swift [hunk 1]
```diff
@@ -40,6 +44,29 @@ extension ValidatorResults.Nested: ValidatorResult {
...
+    public var isFailure: Bool {
+        !self.results.flatMap { $0.value }
+            .filter { $0.isFailure }.isEmpty
+    }
+    
+    public var successDescription: String? {
+        self.results.flatMap { nested in
+            nested.value.filter { !$0.isFailure }
+                .compactMap { $0.successDescription }
+                .map { "[\(nested.key)] \($0)" }
```
Comment (tanner0101) [inline Sources/Vapor/Validation/ValidatorResult.swift]: Maybe something like `Index \(key)` would be more readable?

> Index 1 title contains '€' (allowed: whitespace, A-Z, a-z, 0-9)

What does the full error output here look like?

File: Sources/Vapor/Validation/Validations.swift [hunk 2]
```diff
@@ -34,6 +34,17 @@ public struct Validations {
         self.storage.append(validation)
     }
     
+    public mutating func addForEach(
+        _ key: ValidationKey,
```
Comment (tanner0101) [inline Sources/Vapor/Validation/Validations.swift]: What about `add(forEach:)`?

File: Sources/Vapor/Validation/ValidatorResult.swift [hunk 3]
```diff
@@ -2,6 +2,10 @@ public struct ValidatorResults {
     public struct Nested {
         public let results: [ValidatorResult]
     }
+
+    public struct NestedCollection {
+        public let results: [Int: [ValidatorResult]]
```
Comment (tanner0101) [inline Sources/Vapor/Validation/ValidatorResult.swift]: Could this be `[[ValidatorResult]]`?

File: Sources/Vapor/Validation/ValidatorResult.swift [hunk 4]
```diff
@@ -40,6 +44,29 @@ extension ValidatorResults.Nested: ValidatorResult {
...
 }
 
+extension ValidatorResults.NestedCollection: ValidatorResult {
+    public var isFailure: Bool {
+        !self.results.flatMap { $0.value }
+            .filter { $0.isFailure }.isEmpty
+    }
+    
+    public var successDescription: String? {
+        self.results.flatMap { nested in
```
Comment (tanner0101) [inline Sources/Vapor/Validation/ValidatorResult.swift]: With `[[ValidatorResult]]` you could do `self.results.enumerated()` to get the key.

File: Tests/VaporTests/ValidationTests.swift [hunk 5]
```diff
@@ -354,5 +431,11 @@ private final class User: Validatable, Codable {
             pet.add("age", as: Int.self, is: .range(3...))
         }
         v.add("isAdmin", as: Bool.self)
+        // validate arrays of hobbies
+        v.add("hobbies", as: [Hobby].self, is: !.empty)
```
Comment (tanner0101) [inline Tests/VaporTests/ValidationTests.swift]: This is going to decode `[Hobby]` which will fail with a Codable error if any of the data is malformed. I thinking putting this _after_ the `v.add(forEach:)` will resolve that issue, but we should have a test for it.

File: Sources/Vapor/Validation/ValidatorResult.swift (hunk 1 above)
Comment (simonedelmann) [reply Sources/Vapor/Validation/ValidatorResult.swift]: Changed!

File: Sources/Vapor/Validation/Validations.swift (hunk 2 above)
Comment (simonedelmann) [reply Sources/Vapor/Validation/Validations.swift]: Like so?
```swift
v.add(forEach: "hobbies") { hobby in
  // ...
}
```

File: Tests/VaporTests/ValidationTests.swift (hunk 5 above)
Comment (simonedelmann) [reply Tests/VaporTests/ValidationTests.swift]: IMO this is no issue, as the validations will not stop running if one of them fails. For the user it makes no difference if there is a codable error thrown before or after the nested validation errors. 
Although it the validation for type `[Hobby].self` is useless (regarding decoding) if the nested validations are set properly.

File: Tests/VaporTests/ValidationTests.swift [hunk 6]
```diff
@@ -354,5 +431,11 @@ private final class User: Validatable, Codable {
             pet.add("age", as: Int.self, is: .range(3...))
         }
         v.add("isAdmin", as: Bool.self)
+        // validate arrays of hobbies
+        v.add(forEach: "hobbies") { hobby in
+            hobby.add("title", as: String.self,
+                    is: .count(5...) && .characterSet(.alphanumerics + .whitespaces))
```
Comment (tanner0101) [inline Tests/VaporTests/ValidationTests.swift]: Xcode indentation here, please use:
```
foo(
    bar: ...,
    baz: ...
)
```

Comment (simonedelmann) [reply Tests/VaporTests/ValidationTests.swift]: Fixed! I also fixed the `Pet` validation above.

**Comment by simonedelmann** [comment]
I've added the whole error message in the tests and also in the PR description above.

**Comment by simonedelmann** [comment]
If https://github.com/vapor/vapor/pull/2241 will get merged before this PR, we will need to add support for nested collections within `ValidationsError`.

**Comment by piers12** [comment]
Any update on this PR? would be great to get this merged in.

**Comment by simonedelmann** [comment]
Unfortunately no. This PR maybe needs some updates, as it was written while Vapor 4 still was in beta. I am not aware of the changes since then. Feel free to investigate!

**Comment by tanner0101** [comment]
Merging this into a local branch for further development. Thanks @simonedelmann !

**Pull request merged**